src/charm.py

```diff
@@ -738,10 +841,105 @@ def _get_initialization_secret_from_peer_relation(self) -> Tuple[str, List[str]]
         except (TypeError, SecretNotFoundError, AttributeError):
             raise PeerSecretError(secret_name=VAULT_INITIALIZATION_SECRET_LABEL)
 
+    def _create_s3_session(self, s3_parameters: dict) -> Optional[boto3.session.Session]:
+        """Creates S3 session.
+
+        Args:
+            s3_parameters: Dictionary of the S3 parameters.
+
+        Returns:
+            boto3.session.Session: S3 session.
+        """
+        try:
+            session = boto3.session.Session(
+                aws_access_key_id=s3_parameters["access-key"],
+                aws_secret_access_key=s3_parameters["secret-key"],
+                region_name=s3_parameters["region"],
+            )
+            session.resource("s3", endpoint_url=s3_parameters["endpoint"])
+            return session
+        except KeyError as e:
+            logger.warning("Missing required S3 parameter: %s", e)
+            return None
+        except ValueError as e:
+            logger.warning("Error creating resource: %s", e)
+            return None
+
+    def _retrieve_s3_parameters(self) -> Tuple[Dict, List[str]]:
+        """Retrieve S3 parameters from the S3 integrator relation."""
+        s3_parameters = self.s3.get_s3_connection_info()
+        required_parameters = [
+            "bucket",
+            "access-key",
+            "secret-key",
+            "endpoint",
+        ]
+        missing_required_parameters = [
+            param for param in required_parameters if param not in s3_parameters
+        ]
+        if missing_required_parameters:
+            logger.warning(
+                "Missing required S3 parameters in relation with S3 integrator: %s",
+                missing_required_parameters,
+            )
+            return {}, missing_required_parameters
+        if "region" not in s3_parameters:
+            s3_parameters["region"] = "us-east-1"
+        for key, value in s3_parameters.items():
+            if isinstance(value, str):
+                s3_parameters[key] = value.strip()
+
+        return s3_parameters, []
+
+    def _create_s3_bucket(
+        self,
+        session: boto3.session.Session,
+        bucket_name: str,
+        endpoint: str,
+        region: str,
+    ) -> None:
+        """Create S3 bucket.
+
+        If the bucket already exists, it will be skipped.
+
+        Args:
+            session: S3 session.
+            region: S3 region.
+            bucket_name: S3 bucket name.
+            endpoint: S3 endpoint.
+        """
+        s3 = session.resource("s3", endpoint_url=endpoint)
+        try:
+            # Checking if bucket already exists
+            bucket = s3.Bucket(bucket_name)
+            bucket.meta.client.head_bucket(Bucket=bucket_name)
+            logger.info("Bucket %s exists.", bucket_name)
+            return
+        except ClientError:
+            logger.warning("Bucket %s doesn't exist, creating it.", bucket_name)
```

Comment:
Why have a warning here? We should have warning in expected contexts, not in normal ones.

Reply:
logging improved.